src/irc/core/irc-cap.c

```diff
@@ -79,73 +79,148 @@ static void cap_emit_signal (IRC_SERVER_REC *server, char *cmd, char *args)
+		for (i = 0; i < caps_length; i++) {
+			char *key, *val;
 
-		/* Request the required caps, if any */
-		if (server->cap_queue == NULL) {
-			cap_finish_negotiation(server);
+			if (!parse_cap_name(caps[i], &key, &val)) {
+				g_warning("Invalid CAP %s key/value pair", evt);
+				continue;
+			}
+
+			if (!g_hash_table_insert(server->cap_supported, key, val)) {
+				/* The specification doesn't say anything about
+				 * duplicated values, let's just warn the user */
+				g_warning("The server sent the %s capability twice", key);
+			}
 		}
-		else {
-			cmd = g_string_new("CAP REQ :");
 
-			avail_caps = 0;
+		/* A multiline response is always terminated by a normal one,
+		 * wait until we receive that one to require any CAP */
+		if (multiline == FALSE) {
+			/* No CAP has been requested */
+			if (server->cap_queue == NULL) {
+				cap_finish_negotiation(server);
+			}
+			else {
+				cmd = g_string_new("CAP REQ :");
+
+				avail_caps = 0;
 
-			/* Check whether the cap is supported by the server */
-			for (tmp = server->cap_queue; tmp != NULL; tmp = tmp->next) {
-				if (gslist_find_string(server->cap_supported, tmp->data)) {
-					if (avail_caps > 0)
-						g_string_append_c(cmd, ' ');
-					g_string_append(cmd, tmp->data);
+				/* Check whether the cap is supported by the server */
+				for (tmp = server->cap_queue; tmp != NULL; tmp = tmp->next) {
+					if (g_hash_table_lookup_extended(server->cap_supported, tmp->data, NULL, NULL)) {
+						if (avail_caps > 0)
+							g_string_append_c(cmd, ' ');
+						g_string_append(cmd, tmp->data);
 
-					avail_caps++;
+						avail_caps++;
+					}
 				}
-			}
 
-			/* Clear the queue here */
-			gslist_free_full(server->cap_queue, (GDestroyNotify) g_free);
-			server->cap_queue = NULL;
+				/* Clear the queue here */
+				gslist_free_full(server->cap_queue, (GDestroyNotify) g_free);
+				server->cap_queue = NULL;
 
-			/* If the server doesn't support any cap we requested close the negotiation here */
-			if (avail_caps > 0)
-				irc_send_cmd_now(server, cmd->str);
-			else
-				cap_finish_negotiation(server);
+				/* If the server doesn't support any cap we requested close the negotiation here */
+				if (avail_caps > 0)
+					irc_send_cmd_now(server, cmd->str);
+				else
+					cap_finish_negotiation(server);
 
-			g_string_free(cmd, TRUE);
+				g_string_free(cmd, TRUE);
+			}
 		}
 	}
-	else if (!g_strcmp0(evt, "ACK")) {
+	else if (!g_ascii_strcasecmp(evt, "ACK")) {
 		int got_sasl = FALSE;
 
 		/* Emit a signal for every ack'd cap */
 		for (i = 0; i < caps_length; i++) {
 			disable = (*caps[i] == '-');
 
 			if (disable)
-				server->cap_active = gslist_remove_string(server->cap_active, caps[i] + 1);
+				server->cap_active = gslist_delete_string(server->cap_active, caps[i] + 1, g_free);
 			else
 				server->cap_active = g_slist_prepend(server->cap_active, g_strdup(caps[i]));
 
-			if (!g_strcmp0(caps[i], "sasl"))
+			if (!strcmp(caps[i], "sasl"))
```

Comment:
This change is just increasing the crashiness lol.

Reply:
Not quite as the right hand side is a constant string and the left hand side comes from a null-terminated vector of strings, I used the plain `strcmp` because of that.

Reply:
It just seemed pointless to bother changing it, but sure, it's harmless.